### PR TITLE
Fix rz_name_validate_char in name.c & update test

### LINKED ISSUE #400: Preserve valid names for symbols and sections
| Questions                                            | Answers
|------------------------------------------------------|--------------------
| OS/arch/bits (mandatory)                             | -
| File format of the file you reverse (mandatory)      | ELF
| Architecture/bits of the file (mandatory)            | ARM
| rizin -v full output, **not truncated** (mandatory)  | rizin 0.1.0-git 26825 @ linux-x86-64 git.4.5.1-859-gd17c2574d
commit: d17c2574d8ef762c351d65fde8230f0cc5f117c0 build: 2021-01-20__13:20:57

### Expected behavior

See the `readelf` output, for example:
```
[i] ℤ readelf -SW 006.make.elf.arm32                                                                                                                                                                                              15:16:08 
There are 28 section headers, starting at offset 0x2caa0:

Section Headers:
  [Nr] Name              Type            Addr     Off    Size   ES Flg Lk Inf Al
  [ 0]                   NULL            00000000 000000 000000 00      0   0  0
  [ 1] .interp           PROGBITS        00010154 000154 000019 00   A  0   0  1
  [ 2] .note.ABI-tag     NOTE            00010170 000170 000020 00   A  0   0  4
  [ 3] .note.gnu.build-id NOTE            00010190 000190 000024 00   A  0   0  4
  [ 4] .gnu.hash         GNU_HASH        000101b4 0001b4 000cbc 04   A  5   0  4
  [ 5] .dynsym           DYNSYM          00010e70 000e70 001a80 10   A  6   1  4
  [ 6] .dynstr           STRTAB          000128f0 0028f0 0015b6 00   A  0   0  1
  [ 7] .gnu.version      VERSYM          00013ea6 003ea6 000350 02   A  5   0  2
  [ 8] .gnu.version_r    VERNEED         000141f8 0041f8 000090 00   A  6   3  4
  [ 9] .rel.dyn          REL             00014288 004288 000048 08   A  5   0  4
  [10] .rel.plt          REL             000142d0 0042d0 0003e8 08  AI  5  22  4
  [11] .init             PROGBITS        000146b8 0046b8 00000c 00  AX  0   0  4
  [12] .plt              PROGBITS        000146c4 0046c4 0005f0 04  AX  0   0  4
  [13] .text             PROGBITS        00014cb8 004cb8 021060 00  AX  0   0  8
  [14] .fini             PROGBITS        00035d18 025d18 000008 00  AX  0   0  4
  [15] .rodata           PROGBITS        00035d20 025d20 0058d8 00   A  0   0  4
  [16] .ARM.exidx        ARM_EXIDX       0003b5f8 02b5f8 000008 00  AL 13   0  4
  [17] .eh_frame         PROGBITS        0003b600 02b600 000004 00   A  0   0  4
  [18] .init_array       INIT_ARRAY      0004bef8 02bef8 000004 04  WA  0   0  4
  [19] .fini_array       FINI_ARRAY      0004befc 02befc 000004 04  WA  0   0  4
  [20] .data.rel.ro      PROGBITS        0004bf00 02bf00 000004 00  WA  0   0  8
  [21] .dynamic          DYNAMIC         0004bf04 02bf04 0000f8 08  WA  6   0  4
  [22] .got              PROGBITS        0004c000 02c000 000204 04  WA  0   0  4
  [23] .data             PROGBITS        0004c208 02c208 00072c 00  WA  0   0  8
  [24] .bss              NOBITS          0004c938 02c934 002dd8 00  WA  0   0  8
  [25] .ARM.attributes   ARM_ATTRIBUTES  00000000 02c934 00002f 00      0   0  1
  [26] .gnu_debuglink    PROGBITS        00000000 02c964 000034 00      0   0  4
  [27] .shstrtab         STRTAB          00000000 02c998 000108 00      0   0  1
Key to Flags:
  W (write), A (alloc), X (execute), M (merge), S (strings), I (info),
  L (link order), O (extra OS processing required), G (group), T (TLS),
  C (compressed), x (unknown), o (OS specific), E (exclude),
  y (purecode), p (processor specific)
```

### Actual behavior

It substitutes characters like `-` in the section name: `.note.gnu.build_id` is in fact should be `note.gnu.build-id`, `.note.ABI_tag` should be `.note.ABI-tag`:
```
[i] ℤ rizin 006.make.elf.arm32                                                                                                                                                                                                    
[0x00016d10]> iS
[Sections]

nth paddr          size vaddr         vsize perm name
―――――――――――――――――――――――――――――――――――――――――――――――――――――
0   0x00000000      0x0 0x00000000      0x0 ---- 
1   0x00000154     0x19 0x00010154     0x19 -r-- .interp
2   0x00000170     0x20 0x00010170     0x20 -r-- .note.ABI_tag
3   0x00000190     0x24 0x00010190     0x24 -r-- .note.gnu.build_id
4   0x000001b4    0xcbc 0x000101b4    0xcbc -r-- .gnu.hash
5   0x00000e70   0x1a80 0x00010e70   0x1a80 -r-- .dynsym
6   0x000028f0   0x15b6 0x000128f0   0x15b6 -r-- .dynstr
7   0x00003ea6    0x350 0x00013ea6    0x350 -r-- .gnu.version
8   0x000041f8     0x90 0x000141f8     0x90 -r-- .gnu.version_r
9   0x00004288     0x48 0x00014288     0x48 -r-- .rel.dyn
10  0x000042d0    0x3e8 0x000142d0    0x3e8 -r-- .rel.plt
11  0x000046b8      0xc 0x000146b8      0xc -r-x .init
12  0x000046c4    0x5f0 0x000146c4    0x5f0 -r-x .plt
13  0x00004cb8  0x21060 0x00014cb8  0x21060 -r-x .text
14  0x00025d18      0x8 0x00035d18      0x8 -r-x .fini
15  0x00025d20   0x58d8 0x00035d20   0x58d8 -r-- .rodata
16  0x0002b5f8      0x8 0x0003b5f8      0x8 -r-- .ARM.exidx
17  0x0002b600      0x4 0x0003b600      0x4 -r-- .eh_frame
18  0x0002bef8      0x4 0x0004bef8      0x4 -rw- .init_array
19  0x0002befc      0x4 0x0004befc      0x4 -rw- .fini_array
20  0x0002bf00      0x4 0x0004bf00      0x4 -rw- .data.rel.ro
21  0x0002bf04     0xf8 0x0004bf04     0xf8 -rw- .dynamic
22  0x0002c000    0x204 0x0004c000    0x204 -rw- .got
23  0x0002c208    0x72c 0x0004c208    0x72c -rw- .data
24  0x0002c934      0x0 0x0004c938   0x2dd8 -rw- .bss
25  0x0002c934     0x2f 0x00000000     0x2f ---- .ARM.attributes
26  0x0002c964     0x34 0x00000000     0x34 ---- .gnu_debuglink
27  0x0002c998    0x108 0x00000000    0x108 ---- .shstrtab
```

### Steps to reproduce the behavior 

```
$ rizin 006.make.elf.arm32
[0x00016d10]> iS
```

See implementation in:
- `librz/bin/formats/elf`
- `librz/bin/p/bin_elf.c`

### Additional Logs, screenshots, source-code,  configuration dump, ...

[006.make.elf.arm32.zip](https://github.com/rizinorg/rizin/files/5840708/006.make.elf.arm32.zip)

### PR DESCRIPTION
**Pull request by Yuuoniy**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

...

**Test plan**

...

**Closing issues**

Closes https://github.com/rizinorg/rizin/issues/400

### CODE REVIEW

**Comment by XVilka** [comment]
@Yuuoniy looks like it broke some unrelated tests https://builds.sr.ht/~xvilka/job/453813
Basically, because some names still need `-` character to be filtered out, while some - don't.

**Comment by kazarmy** [comment]
@ret2libc Will rizin confuse `-` in names with subtraction?

**Comment by thestr4ng3r** [comment]
I think we should strive to support `-` in names, but waiting for @ret2libc about whether anything speaks against it. For expressions, my idea would be that you would then have to write `"some-name"-42` for example.

**Comment by ret2libc** [comment]
> I think we should strive to support `-` in names, but waiting for @ret2libc about whether anything speaks against it. For expressions, my idea would be that you would then have to write `"some-name"-42` for example.

The expression as you wrote it would not be a problem for the parser, the problem is RzNum. I bet that `rz_num_math` of `"some-name"-42` would not work. I have not tested it though, just guessing.

**Comment by Yuuoniy** [comment]
@XVilka  OK, I would look into these cases to figure out it.
By the way, I found out that other characters like '[' may also be mistakenly filtered out.
case :
```
[XX] db/formats/mach0/coresymbolication Xcode symbols cache (flags)
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'fdj@0x100006430~{:
' bins/mach0/D9A37B67-F10A-35AA-A852-ABFBBECC03AE.symbols
-- stdout
@@ -1,3 +1,3 @@
 offset: 4294992900
-name: sym.__AppDelegate_application:configurationForConnectingSceneSession:options:
+name: sym.-_AppDelegate_application:configurationForConnectingSceneSession:options:
```
I think the expected output should be sym.-[AppDelegate_application:configurationForConnectingSceneSession:options:]

**Comment by Yuuoniy** [comment]
Hi, I am not sure when `-`  character should/shouldn't be filtered out , here are some ideas:
1. For string name: 
It seems rizin use content of string generate name for string.  for example, name of string` "file name is:%@ - protection class:%@"`  is` file_name_is:_____protection_class` . In this case I think `-` character  should be filtered out, or it would look strange.
2. For sections/symbols name:
`-` character  should be preserved. So `.note.ABI-tag`  `.note.gnu.build-id ` `sym.imp.api-ms-win-crt-stdio-l1-1-0.` .. could be shown correctly. Besides, I think `drectve-0 debug_S-1 debug_T-2 text_mn-3..` are also reasonable.
But there is a case I am uncertain:
```
[XX] db/formats/mach0/coresymbolication Xcode symbols cache (flags)
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'fdj@0x100006430~{:
' bins/mach0/D9A37B67-F10A-35AA-A852-ABFBBECC03AE.symbols
-- stdout
@@ -1,3 +1,3 @@
 offset: 4294992900
-name: sym.__AppDelegate_application:configurationForConnectingSceneSession:options:
+name: sym.-_AppDelegate_application:configurationForConnectingSceneSession:options:
```
here the realname should be `-[AppDelegate application:configurationForConnectingSceneSession:options:]`, if we expect rizin to print the realname, '[' ']' should also be preserved, or `sym.__AppDelegate_application:configurationForConnectingSceneSession:options:` output is enough.
How do you think?

**Comment by XVilka** [comment]
@Yuuoniy you are absolutely right in both cases. If it's an Objective C it's better to preserve the square brackets, otherwise - better to drop them.

**Comment by ret2libc** [comment]
One thing that can be a problem with this PR:
How would command handlers handle things like `sym.imp.api-ms-win-crt-stdio-l1-1-0.dll___stdio_common_vfscanf` as argument? Even if newshell allows to wrap it in quotes (e.g. `"sym.imp.api-ms-win-crt-stdio-l1-1-0.dll___stdio_common_vfscanf"`), there are still many oldshell handlers and I'm not sure they would accept `sym.imp.api-ms-win-crt-stdio-l1-1-0.dll___stdio_common_vfscanf` as argument, because of the `-`. Even the `s` command... Will it work? Could you test it?

**Comment by Yuuoniy** [comment]
Oh, it is indeed a problem. Test with `s` command, and it doesn't work.
The key point may be to filter characters correctly for different purposes.  I would try to sort it out.

**Comment by ret2libc** [comment]
I think section names as showed in `iS` can be kept with `-`, etc., but flag names should, at least for now, use a more aggressive filtering as currently done.

**Comment by Yuuoniy** [comment]
I add a flag `strict` to decide whether use a more strict filtering or not.  For most cases, a strict mode is required.

**Review by XVilka — APPROVED** [review]

**Comment by XVilka** [comment]
@Yuuoniy by the way, please in the future start commit messages from the capital letter, you can read more about our development/contributor guidelines here:
- https://github.com/rizinorg/rizin/blob/dev/CONTRIBUTING.md#commit-message-rules
- https://github.com/rizinorg/rizin/blob/dev/DEVELOPERS.md

Also a general recommendation is to use the different branch name for every PR, so it can be rebased/merged easier.

**Comment by Yuuoniy** [comment]
@XVilka  Thanks for your reminder, I'll pay attention next time.

File: test/db/formats/mach0/coresymbolication [hunk 1]
```diff
@@ -71,15 +71,15 @@ CMDS=<<EOF
...
-"f sym.__SceneDelegate_scene:willConnectToSession:options: 4 0x100006514"
-"f sym.__SceneDelegate_sceneDidDisconnect: 4 0x100006518"
-"f sym.__SceneDelegate_sceneDidBecomeActive: 4 0x10000651c"
-"f sym.__SceneDelegate_sceneWillResignActive: 4 0x100006520"
-"f sym.__SceneDelegate_sceneWillEnterForeground: 4 0x100006524"
-"f sym.__SceneDelegate_sceneDidEnterBackground: 4 0x100006528"
-"f sym.__SceneDelegate_window 16 0x10000652c"
-"f sym.__SceneDelegate_setWindow: 20 0x10000653c"
-"f sym.__SceneDelegate_.cxx_destruct 20 0x100006550"
+"f sym.-[SceneDelegate_scene:willConnectToSession:options:] 4 0x100006514"
```
Comment (ret2libc) [inline test/db/formats/mach0/coresymbolication]: does this work properly?

**Review by ret2libc — COMMENTED** [review]
I think this can be a good compromise for a short/mid-term solution. However please check if `?v sym.-[SceneDelegate_scene:willConnectToSession:options:]` or `?v sym.-[SceneDelegate_scene:willConnectToSession:options:] + 0x100` still work.

In general, the change is just in one function, so even if we have to change this in the future it will be easy. Thanks :)

File: test/db/formats/mach0/coresymbolication (hunk 1 above)
Comment (XVilka) [reply test/db/formats/mach0/coresymbolication]:
```
[i] ℤ rizin /bin/ls
[0x00006b60]> "f bla-[sdgsf] 4 @ 0x6b60"
[0x00006b60]> f~bal
[0x00006b60]> f~bla
0x00000000 4 bla__sdgsf
```
Then if we check it in `fj` output:
```json
  {
    "name": "bla__sdgsf",
    "size": 4,
    "offset": 0,
    "comment": "0x6b60"
  },
```

Comment (XVilka) [reply test/db/formats/mach0/coresymbolication]: So probably we should wait for this particular change until `f` ported to the newshell and maybe rest of the commands that work with flags.

**Comment by Yuuoniy** [comment]
OK, I would like to leave character filtering of symbols/flags name the way they are, so that insure other commands still work.  '-' only preserved in sections name at this time.

**Review by wargio — APPROVED** [review]
LGTM. good work!

**Review by XVilka — APPROVED** [review]

**Pull request merged**